File: frontend/src/api/cafeBilling.ts
import { apiRequest } from "./client";

export type CafeBillItem = {
  order_public_id: string;
  order_number: string;
  source_channel: "qr_customer" | "order_taker" | "billing_counter" | "manager";
  order_status: string;
  order_item_id: number;
  menu_item_name: string;
  product_id: number | null;
  sku: string | null;
  quantity: number;
  unit_price: string;
  discount: string;
  line_total: string;
  billed: boolean;
  excluded_reason: string | null;
};

export type CafeBillQuote = {
  source_type: "cafe_table_session" | "cafe_takeaway";
  source_id: string;
  branch_id: number;
  table_session_public_id: string | null;
  table_session_status: string | null;
  source_version: number;
  subtotal: string;
  discount_total: string;
  taxable_total: string;
  cgst_total: string;
  sgst_total: string;
  igst_total: string;
  cess_total: string;
  round_off: string;
  grand_total: string;
  eligible_items: CafeBillItem[];
  excluded_items: CafeBillItem[];
};

export type CafeBillPaymentInput = {
  payment_mode_id?: number | null;
  amount: string;
  reference_number?: string | null;
  notes?: string | null;
};

export type CafeBillRequest = {
  expected_version: number;
  customer_id?: number | null;
  payments: CafeBillPaymentInput[];
};

export type CafeReceipt = {
  invoice_id: number;
  invoice_number: string;
  source_type: string;
  source_id: string;
  cafe_name: string;
  branch_name: string;
  invoice_type: string;
  invoice_status: string;
  payment_status: string;
  issued_at: string | null;
  subtotal: string;
  discount_total: string;
  taxable_total: string;
  cgst_total: string;
  sgst_total: string;
  igst_total: string;
  cess_total: string;
  round_off: string;
  grand_total: string;
  paid_amount: string;
  balance_due: string;
  gstin: null;
  items: Array<{
    name: string;
    sku: string;
    quantity: string;
    unit_price: string;
    discount: string;
    line_total: string;
  }>;
  payments: Array<{
    mode_name: string | null;
    amount: string;
    reference_number: string | null;
    is_credit_marker: boolean;
  }>;
};

export type CafeBillResult = {
  receipt: CafeReceipt;
  table_session_status: string | null;
  order_status: string | null;
  closed: boolean;
  idempotent_replay: boolean;
};

export function quoteCafeTableSession(token: string, publicId: string): Promise<CafeBillQuote> {
  return apiRequest<CafeBillQuote>(`/cafe/billing/table-sessions/${publicId}/quote`, {}, token);
}

export function quoteCafeOrder(token: string, publicId: string): Promise<CafeBillQuote> {
  return apiRequest<CafeBillQuote>(`/cafe/billing/orders/${publicId}/quote`, {}, token);
}

export function billCafeTableSession(
  token: string,
  publicId: string,
  payload: CafeBillRequest,
  idempotencyKey: string,
): Promise<CafeBillResult> {
  return apiRequest<CafeBillResult>(
    `/cafe/billing/table-sessions/${publicId}/bill`,
    {
      method: "POST",
      headers: { "Idempotency-Key": idempotencyKey },
      body: JSON.stringify(payload),
    },
    token,
  );
}

export function billCafeOrder(
  token: string,
  publicId: string,
  payload: CafeBillRequest,
  idempotencyKey: string,
): Promise<CafeBillResult> {
  return apiRequest<CafeBillResult>(
    `/cafe/billing/orders/${publicId}/bill`,
    {
      method: "POST",
      headers: { "Idempotency-Key": idempotencyKey },
      body: JSON.stringify(payload),
    },
    token,
  );
}

export function getCafeReceipt(token: string, invoiceId: number): Promise<CafeReceipt> {
  return apiRequest<CafeReceipt>(`/cafe/billing/invoices/${invoiceId}/receipt`, {}, token);
}

export function collectCafeInvoicePayment(
  token: string,
  invoiceId: number,
  payment: CafeBillPaymentInput,
): Promise<CafeBillResult> {
  return apiRequest<CafeBillResult>(
    `/cafe/billing/invoices/${invoiceId}/payments`,
    { method: "POST", body: JSON.stringify({ payment }) },
    token,
  );
}

File: backend/app/schemas/cafe_billing.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator

from app.models import CafeOrderSource, CafeOrderStatus, InvoicePaymentStatus, InvoiceStatus, TableSessionStatus
from app.schemas.invoices import InvoicePaymentCreate


class CafeBillingPaymentInput(InvoicePaymentCreate):
    pass


class CafeBillRequest(BaseModel):
    expected_version: int | None = Field(default=None, ge=1)
    customer_id: int | None = Field(default=None, ge=1)
    payments: list[CafeBillingPaymentInput] = Field(default_factory=list, max_length=8)


class CafeBillingItemRead(BaseModel):
    order_public_id: str
    order_number: str
    source_channel: CafeOrderSource
    order_status: CafeOrderStatus
    order_item_id: int
    menu_item_name: str
    product_id: int | None
    sku: str | None
    quantity: int
    unit_price: Decimal
    discount: Decimal
    line_total: Decimal
    billed: bool
    excluded_reason: str | None = None


class CafeBillQuoteRead(BaseModel):
    source_type: str
    source_id: str
    branch_id: int
    table_session_public_id: str | None = None
    table_session_status: TableSessionStatus | None = None
    source_version: int
    subtotal: Decimal
    discount_total: Decimal
    taxable_total: Decimal
    cgst_total: Decimal
    sgst_total: Decimal
    igst_total: Decimal
    cess_total: Decimal
    round_off: Decimal
    grand_total: Decimal
    eligible_items: list[CafeBillingItemRead]
    excluded_items: list[CafeBillingItemRead]


class CafeBillPaymentRead(BaseModel):
    mode_name: str | None
    amount: Decimal
    reference_number: str | None
    is_credit_marker: bool


class CafeReceiptItemRead(BaseModel):
    name: str
    sku: str
    quantity: Decimal
    unit_price: Decimal
    discount: Decimal
    line_total: Decimal


class CafeReceiptRead(BaseModel):
    invoice_id: int
    invoice_number: str
    source_type: str
    source_id: str
    cafe_name: str
    branch_name: str
    invoice_type: str
    invoice_status: InvoiceStatus
    payment_status: InvoicePaymentStatus
    issued_at: datetime | None
    subtotal: Decimal
    discount_total: Decimal
    taxable_total: Decimal
    cgst_total: Decimal
    sgst_total: Decimal
    igst_total: Decimal
    cess_total: Decimal
    round_off: Decimal
    grand_total: Decimal
    paid_amount: Decimal
    balance_due: Decimal
    gstin: None = None
    items: list[CafeReceiptItemRead]
    payments: list[CafeBillPaymentRead]


class CafeBillResultRead(BaseModel):
    receipt: CafeReceiptRead
    table_session_status: TableSessionStatus | None = None
    order_status: CafeOrderStatus | None = None
    closed: bool
    idempotent_replay: bool = False


class CafePaymentCollectRequest(BaseModel):
    payment: CafeBillingPaymentInput


class CafeCloseRequest(BaseModel):
    expected_version: int = Field(ge=1)


class CafeBillingSourceRead(BaseModel):
    source_type: str
    source_id: str
    label: str
    branch_id: int
    status: str
    source_version: int
    order_count: int
    grand_total_hint: Decimal

    @field_validator("source_type")
    @classmethod
    def validate_source(cls, value: str) -> str:
        if value not in {"cafe_table_session", "cafe_takeaway"}:
            raise ValueError("Unsupported Cafe billing source.")
        return value
